### PR TITLE
ZF3 component compatibility

### PR DESCRIPTION
**Pull request by weierophinney**
This ~~is a work-in-progress for updating~~ patch updates the module to work with ZF3 components.

The patch does the following:

- Updates the minimum supported PHP version to 5.6.
- Updates all component and module dependencies to the latest known stable version; if v3 releases are available, supports these via `||` constraints.
- Moves the `Module` class to the `src/` tree, where it belongs, and:
  - removes the `getAutoloaderConfig()` method, as it was redundant
  - updates how listener aggregates are attached to event managers
- Updates all factories and abstract factories to be forwards compatible with v3 releases.
- Updates the `Application` class to:
  - catch PHP 7 throwables
  - trigger events in a way that is forwards compatible with zend-eventmanager v3.
- Removes the dependency on rwoverdijk/assetmanager, and adds a suggestion for its upcoming 1.7 version, or for the new zfcampus/zf-asset-manager.

### TODO

- ~~Waiting on a release based on rwoverdijk/assetmanager#193 before merging, as the current releases do not work with v3 components.~~

### CODE REVIEW

**Comment by weierophinney** [comment]
Build is passing, so only assetmanager remains!

**Comment by webimpress** [comment]
@weierophinney why we need assetmanager dependency in that package? As I can see these assets are used in `zf-apigility-skeleton` so maybe better will be to move it there. What do you think?

**Comment by weierophinney** [comment]
> why we need assetmanager dependency in that package? As I can see these assets are used in zf-apigility-skeleton so maybe better will be to move it there. What do you think?

As noted on other threads, the assets are used in the skeleton, as well as in the admin UI and zf-apigility-welcome. While we could potentially move them elsewhere, the idea was to ensure that you can add zf-apigility to an existing Zend Framework MVC project, and have everything "just work"; the current solution is tested and known to accomplish that goal. We can revisit the assets sometime again in the future, but for now, I want to keep stability.

**Comment by weierophinney** [comment]
@zfcampus/community-review-team This PR is now ready for review; would one of you be able to do a look-over, please?

File: composer.json [hunk 1]
```diff
@@ -16,52 +16,62 @@
...
+        "zendframework/zend-loader": "^2.5.1",
+        "zendframework/zend-mvc": "^2.7.9 || ^3.0.2",
+        "zendframework/zend-paginator": "^2.7",
+        "zendframework/zend-servicemanager": "^2.7.6 || ^3.1.1",
+        "zendframework/zend-stdlib": "^2.7.7 || ^3.0.1",
+        "zfcampus/zf-api-problem": "^1.2.1",
+        "zfcampus/zf-apigility-provider": "^1.2",
+        "zfcampus/zf-content-negotiation": "^1.2.1",
+        "zfcampus/zf-content-validation": "^1.3.3",
+        "zfcampus/zf-hal": "^1.4",
```
Comment (webimpress) [inline composer.json]: @weierophinney `zf-hal` v1.4 can't be used with `zend-stdlib`.
Please have a look on https://github.com/zfcampus/zf-hal/pull/149

File: composer.json [hunk 2]
```diff
@@ -16,52 +16,62 @@
...
-        "zfcampus/zf-hal": "~1.0",
-        "zfcampus/zf-mvc-auth": "~1.0",
-        "zfcampus/zf-oauth2": "~1.0",
-        "zfcampus/zf-rest": "~1.0",
-        "zfcampus/zf-rpc": "~1.0",
-        "zfcampus/zf-versioning": "~1.0"
+        "php": "^5.6 || ^7.0",
+        "zendframework/zend-db": "^2.8.1",
+        "zendframework/zend-eventmanager": "^2.6.3 || ^3.0.1",
+        "zendframework/zend-loader": "^2.5.1",
```
Comment (webimpress) [inline composer.json]: do we need it here? before it was in `require-dev` section

File: config/module.config.php [hunk 3]
```diff
@@ -1,49 +1,55 @@
...
-            'zf-apigility' => array(
-                'type'  => 'Zend\Mvc\Router\Http\Literal',
-                'options' => array(
+            ],
+        ],
+    ],
+    'router' => [
+        'routes' => [
+            'zf-apigility' => [
+                'type'  => 'literal',
```
Comment (webimpress) [inline config/module.config.php]: are we gonna use here `::class` notation?

one space before `=>` please

File: src/Application.php [hunk 4]
```diff
@@ -52,19 +53,39 @@ public function run()
...
+            if ($response instanceof ResponseInterface) {
+                $event->setName(MvcEvent::EVENT_FINISH);
+                $event->setTarget($this);
+                $event->setResponse($response);
+                $this->response = $response;
+                $events->triggerEvent($event);
+                return $this;
+            }
+
+            return $this->completeRequest($event);
```
Comment (webimpress) [inline src/Application.php]: can we extract this part to separate method to avoid code duplication? is it exactly the same in `catch (Exception $e)` ?

File: src/DbConnectedResourceAbstractFactory.php [hunk 5]
```diff
@@ -99,10 +161,21 @@ protected function getIdentifierFromConfig(array $config)
...
+     *
+     * @param array $config
+     * @param string $requestedName
+     * @return string
+     * @throws ServiceNotCreatedException if the discovered collection class
+     *     does not exist.
+     */
     protected function getCollectionFromConfig(array $config, $requestedName)
     {
         $collection = isset($config['collection_class']) ? $config['collection_class'] : 'Zend\Paginator\Paginator';
```
Comment (webimpress) [inline src/DbConnectedResourceAbstractFactory.php]: `::class` notation please

File: src/DbConnectedResourceAbstractFactory.php [hunk 6]
```diff
@@ -112,15 +185,23 @@ protected function getCollectionFromConfig(array $config, $requestedName)
...
     {
-        $defaultClass  = __NAMESPACE__ . '\DbConnectedResource';
+        $defaultClass  = DbConnectedResource::class;
         $resourceClass = isset($config['resource_class']) ? $config['resource_class'] : $defaultClass;
         if ($resourceClass !== $defaultClass
-            && (
-                !class_exists($resourceClass)
-                || !is_subclass_of($resourceClass, $defaultClass)
-            )
+            && ! (class_exists($resourceClass) || is_subclass_of($resourceClass, $defaultClass))
```
Comment (webimpress) [inline src/DbConnectedResourceAbstractFactory.php]: I think this change should be reverted, because it tries evaluate first condition inside brackets, so:
`class_exists($resourceClass) || is_subclass_of($resourceClass, $defaultClass)`
and in case class does not exist it will try evaluate `is_subclass_of`. Before it was not needed - before if class does not exists there was no check for `is_subclass_of`

File: src/TableGatewayAbstractFactory.php [hunk 7]
```diff
@@ -60,57 +92,120 @@ public function createServiceWithName(ServiceLocatorInterface $services, $name,
...
+     * retrieved from the HydratorManager; otherwise ArraySerializable
+     * will be retrieved.
+     *
+     * @param array $config
+     * @param ContainerInterface $container
+     * @return \Zend\Hydrator\HydratorInterface
+     */
+    protected function getHydratorFromConfig(array $config, ContainerInterface $container)
     {
         $hydratorName = isset($config['hydrator_name']) ? $config['hydrator_name'] : 'ArraySerializable';
```
Comment (webimpress) [inline src/TableGatewayAbstractFactory.php]: can we use `::class` notation here?

File: test/DbConnectedResourceAbstractFactoryTest.php [hunk 8]
```diff
@@ -91,29 +109,31 @@ public function testWillCreateServiceIfDbConnectedSegmentIsValid($configForDbCon
                 'Foo' => $configForDbConnected,
             ],
         ]];
-        $this->services->set('Config', $config);
-        $this->services->set($tableServiceName, new stdClass());
-        $this->assertTrue($this->factory->canCreateServiceWithName($this->services, 'foo', 'Foo'));
+        $this->services->has('config')->willReturn(true);
+        $this->services->get('config')->willReturn($config);
+
```
Comment (webimpress) [inline test/DbConnectedResourceAbstractFactoryTest.php]: why the empty line is here? line below belongs to "arrange" tests part, so better it the empty line will be before "assert" part

File: test/DbConnectedResourceTest.php [hunk 9]
```diff
@@ -1,24 +1,24 @@
...
 {
     public function setUp()
     {
-        $this->table = $this->getMockBuilder('Zend\Db\TableGateway\TableGateway')
-            ->disableOriginalConstructor()
-            ->getMock();
-
-        $this->resource = new DbConnectedResource($this->table, 'id', 'ArrayObject');
+        $this->table    = $this->prophesize(TableGateway::class);
+        $this->resource = new DbConnectedResource($this->table->reveal(), 'id', 'ArrayObject');
```
Comment (webimpress) [inline test/DbConnectedResourceTest.php]: `::class` notation please

File: composer.json (hunk 2 above)
Comment (weierophinney) [inline composer.json]: It's required by `ZF\Apigility\Autoloader`, which is the default autoloader used in  `Module::getAutoloaderConfig()` implementations generated by the admin API. Since zend-loader is no longer a requirement of zend-mvc, we need to make it a requirement here going forward to ensure it's present.

File: config/module.config.php (hunk 3 above)
Comment (weierophinney) [inline config/module.config.php]: Using `::class` notation here is a no-go at this time, due to the fact that we cannot know which of zend-mvc v2 or zend-router v{any} are present. As such, I updated to use the "short name" aliases that are present in both.

I'll fix the space issues shortly.

File: test/TableGatewayAbstractFactoryTest.php [hunk 10]
```diff
@@ -6,99 +6,125 @@
...
-        $this->services->set('Config', ['zf-apigility' => ['db-connected' => ['Foo' => [
-            'table_name'   => 'test',
-        ]]]]);
-
-        $this->services->set('Zend\Db\Adapter\Adapter', new stdClass());
-        $this->assertTrue($this->factory->canCreateServiceWithName($this->services, 'footable', 'Foo\Table'));
+        $this->services->has('config')->willReturn(true);
+        $this->services->get('config')
+           ->willReturn(['zf-apigility' => ['db-connected' => ['Foo' => [
+                'table_name'   => 'test',
```
Comment (webimpress) [inline test/TableGatewayAbstractFactoryTest.php]: one space before `=>`, is it to long to fit one line ?

File: config/module.config.php (hunk 3 above)
Comment (webimpress) [inline config/module.config.php]: you have right, thanks

File: src/TableGatewayAbstractFactory.php (hunk 7 above)
Comment (weierophinney) [inline src/TableGatewayAbstractFactory.php]: No, because the class could be either `Zend\Stdlib\Hydrator\ArraySerializable` or `Zend\Hydrator\ArraySerializable`.

File: composer.json (hunk 1 above)
Comment (weierophinney) [inline composer.json]: I'll update this after doing another zf-hal release; thanks for the heads-up!

File: src/Application.php [hunk 11]
```diff
@@ -105,4 +100,32 @@ public function run()
 
         return $this;
```
Comment (webimpress) [inline src/Application.php]: Can we change here to have:

`return $this->completeRequest($event);`

I saw it somewhere in another PR, but don't know where now.

File: src/TableGatewayAbstractFactory.php [hunk 12]
```diff
@@ -60,57 +92,120 @@ public function createServiceWithName(ServiceLocatorInterface $services, $name,
...
+     *
+     * If the configuration defines an `adapter_name` and a matching service
+     * is discovered, that will be returned.
+     *
+     * If the Adapter service is present, that will be returned (zend-mvc v2).
+     *
+     * Otherwise, the AdapterInterface service is returned.
+     *
+     * @param array $config
+     * @param ContainerInterface $container
```
Comment (webimpress) [inline src/TableGatewayAbstractFactory.php]: missing `@return` tag in php-doc

File: test/ApplicationTest.php [hunk 13]
```diff
@@ -19,37 +22,52 @@ public function setUp()
...
-    public function setUpServices($services, $events, $request, $response)
+    /**
+     * Create and return an Application instance.
+     *
+     * Checks to see which version of zend-mvc is present, and uses that to
+     * determine how to construct the instance.
+     *
+     * @return Application
+     */
+    public function createApplication($services, $events, $request, $response)
```
Comment (webimpress) [inline test/ApplicationTest.php]: missing params in php-doc

File: src/DbConnectedResource.php [hunk 14]
```diff
@@ -1,30 +1,51 @@
...
+     */
     protected $collectionClass;
 
+    /**
+     * @var string Name of identifier field
+     */
     protected $identifierName;
 
+    /**
+     * @var TableGateway
```
Comment (webimpress) [inline src/DbConnectedResource.php]: I think it should be: `Zend\Db\TableGateway\TableGateway` because `TableGatewayPaginator` requires this type (line 119 below)

File: src/Module.php [hunk 15]
```diff
@@ -1,59 +1,65 @@
...
  * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
- * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
+ * @copyright Copyright (c) 2014-2016 Zend Technologies USA Inc. (http://www.zend.com)
  */
 
 namespace ZF\Apigility;
 
 use Zend\Mvc\MvcEvent;
 use Zend\View\Model\JsonModel;
+use ZF\ApiProblem\RenderErrorListener;
```
Comment (webimpress) [inline src/Module.php]: hm... shouldn't it be `ZF\ApiProblem\Listener\RenderErrorListener` ?

File: test/DbConnectedResourceAbstractFactoryTest.php [hunk 16]
```diff
@@ -6,49 +6,59 @@
 
 namespace ZFTest\Apigility;
 
+use Interop\Container\ContainerInterface;
 use PHPUnit_Framework_TestCase as TestCase;
 use stdClass;
```
Comment (webimpress) [inline test/DbConnectedResourceAbstractFactoryTest.php]: it is never used in that class

File: test/TableGatewayAbstractFactoryTest.php [hunk 17]
```diff
@@ -6,99 +6,135 @@
 
 namespace ZFTest\Apigility;
 
+use Interop\Container\ContainerInterface;
 use PHPUnit_Framework_TestCase as TestCase;
 use stdClass;
```
Comment (webimpress) [inline test/TableGatewayAbstractFactoryTest.php]: it is never used in that class

File: test/ApplicationTest.php [hunk 18]
```diff
@@ -70,25 +88,23 @@ public function testRouteListenerRaisingExceptionTriggersDispatchErrorAndSkipsDi
     {
         $phpunit  = $this;
```
Comment (webimpress) [inline test/ApplicationTest.php]: can be removed now, not used

**Comment by webimpress** [comment]
@weierophinney please have a look on my PR to your branch: https://github.com/weierophinney/zf-apigility/pull/5
It fixes most of the above comments.

File: src/Application.php (hunk 11 above)
Comment (weierophinney) [inline src/Application.php]: That would potentially change behavior.

This is verbatim what was already in the exception handler, and I'm not comfortable changing it as it could potentially break applications (double rendering, etc.).

File: src/DbConnectedResource.php (hunk 14 above)
Comment (weierophinney) [inline src/DbConnectedResource.php]: Ooof, that's a hard call. Changing that would be a BC break, as it's tightening the type hint (vs loosening); simultaneously, you are correct in the assertion that the `DbTableGateway` adapter requires the more specific type.

What we should likely do is update the `DbTableGateway` to accept `TableGatewayInterface` instead (it predates that interface, which is why it typehints against the concrete instance).

I'm going to leave this as-is for now.

File: src/Module.php (hunk 15 above)
Comment (weierophinney) [inline src/Module.php]: Yes; good catch!

File: test/DbConnectedResourceAbstractFactoryTest.php (hunk 16 above)
Comment (weierophinney) [inline test/DbConnectedResourceAbstractFactoryTest.php]: Yes it is - line 20 (we mock it for the servcie container).

Comment (weierophinney) [inline test/DbConnectedResourceAbstractFactoryTest.php]: Nevermind, I thought you were pointing out the `ContainerInterface`, not `stdClass`. I'll update. (It was used before I refactored the tests.)

File: src/Application.php (hunk 11 above)
Comment (webimpress) [inline src/Application.php]: hm... why? Isn't `completeRequest` return `$this`? You have the same in `handleException` so I think it should be fine. Can you think about edge case when it will be an issue? Any test case?

**Comment by weierophinney** [comment]
> please have a look on my PR to your branch

argh! I just got to that comment, after fixing the various items you noted!

File: test/DbConnectedResourceAbstractFactoryTest.php (hunk 16 above)
Comment (webimpress) [inline test/DbConnectedResourceAbstractFactoryTest.php]: @weierophinney Please check my PR https://github.com/weierophinney/zf-apigility/pull/5 :)

File: src/Application.php (hunk 11 above)
Comment (weierophinney) [inline src/Application.php]: As noted in weierophinney/zf-apigility#5, I mistook this as being in the exception handler. On review, this is in `run()`, and your suggestion would replace two lines with a single call.

File: src/TableGatewayAbstractFactory.php (hunk 12 above)
Comment (weierophinney) [inline src/TableGatewayAbstractFactory.php]: Added now (`@return AdapterInterface`).

File: test/ApplicationTest.php [hunk 19]
```diff
@@ -19,37 +22,56 @@ public function setUp()
...
-            ->disableOriginalConstructor()
-            ->getMock();
+        $request = $this->prophesize(PhpEnvironment\Request::class);
+        $response = $this->prophesize(PhpEnvironment\Response::class);
 
-        $services = $this->getMockBuilder('Zend\ServiceManager\ServiceManager')
-            ->disableOriginalConstructor()
-            ->getMock();
+        $services = $this->prophesize(ServiceManager::class);
         $this->services = $services = $this->setUpServices($services, $events, $request, $response);
```
Comment (webimpress) [inline test/ApplicationTest.php]: hm... we don't need `$services` variable here, confusing a bit

File: test/ApplicationTest.php [hunk 20]
```diff
@@ -19,37 +22,56 @@ public function setUp()
...
+    /**
+     * Create and return an Application instance.
+     *
+     * Checks to see which version of zend-mvc is present, and uses that to
+     * determine how to construct the instance.
+     *
+     * @param ServiceManager $services
+     * @param EventManager $events
+     * @param PhpEnvironment\Request $request
+     * @param PhpEnvironment\Response $response
```
Comment (webimpress) [inline test/ApplicationTest.php]: `$services`, `$request` and `$response` (should) have type `ObjectProphecy`

Comment (weierophinney) [inline test/ApplicationTest.php]: I'll pass the results of `reveal()` in each case, then; I think it's more important to detail what the end result of what we're passing.

Comment (webimpress) [inline test/ApplicationTest.php]: So maybe both? `ObjectProphecy|ServiceManager` etc

It's not confusing, because you can't pass just `ServiceManager`.

**Comment by weierophinney** [comment]
Thanks for the thorough review, @webimpress !

I'll give this overnight for other folks to look at, and then I think this will be ready to tag, leaving only the admin and skeleton to go!
